File: src/services/kubectl.ts
/**
 * Copyright 2021 Silicon Hills LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import YAML from 'yaml';
import { ExecaChildProcess, Options } from 'execa';
import { KubernetesObject } from '@kubernetes/client-node';
import { Readable } from 'stream';
import CommandService, { RunCallback } from './command';

export default class KubectlService extends CommandService {
  command = 'kubectl';

  private getStdin(
    stdin: string | KubernetesObject | KubernetesObject[]
  ): string {
    if (Array.isArray(stdin)) {
      return this.resources2String(stdin);
    }
    if (typeof stdin === 'object') {
      return this.resources2String([stdin]);
    }
    return stdin as string;
  }

  async help(options?: Options, cb?: RunCallback) {
    return this.run('--help', options, cb);
  }

  async apply(
    applyOptions: Partial<ApplyOptions> | string = {},
    options?: Options
  ) {
    let { stdin } = applyOptions as Partial<GetOptions>;
    if (typeof applyOptions === 'string') stdin = applyOptions;
    const { file, stdout } = {
      ...((stdin ? { file: '-' } : {}) as Partial<GetOptions>),
      ...(typeof applyOptions === 'string'
        ? ({} as Partial<GetOptions>)
        : applyOptions)
    };
    return this.run(
      ['apply', ...(file ? ['-f', file] : [])],
      options,
      (p: ExecaChildProcess) => {
        if (stdin) {
          const stream = Readable.from([this.getStdin(stdin)]);
          if (p.stdin) stream.pipe(p.stdin);
        }
        if (stdout) {
          p.stderr?.pipe(process.stderr);
          p.stdout?.pipe(process.stdout);
        }
      }
    );
  }

  async get<T = any>(
    getOptions: Partial<GetOptions> | string = {},
    options?: Options
  ): Promise<T> {
    const { stdout } = getOptions as Partial<GetOptions>;
    let { stdin } = getOptions as Partial<GetOptions>;
    if (typeof getOptions === 'string') stdin = getOptions;
    const { file, output, ignoreNotFound } = {
      ignoreNotFound: true,
      ...((stdin ? { file: '-' } : {}) as Partial<GetOptions>),
      ...(typeof getOptions === 'string'
        ? ({} as Partial<GetOptions>)
        : getOptions)
    };
    return this.run<T>(
      [
        'get',
        ...(file ? ['-f', file] : []),
        ...(output ? ['-o', output] : []),
        ...(ignoreNotFound ? ['--ignore-not-found'] : [])
      ],
      options,
      (p: ExecaChildProcess) => {
        if (stdin) {
          const stream = Readable.from([this.getStdin(stdin)]);
          if (p.stdin) stream.pipe(p.stdin);
        }
        if (stdout) {
          p.stderr?.pipe(process.stderr);
          p.stdout?.pipe(process.stdout);
        }
      }
    );
  }

  async delete(
    deleteOptions: Partial<DeleteOptions> | string = {},
    options?: Options
  ) {
    let { stdin } = deleteOptions as Partial<GetOptions>;
    if (typeof deleteOptions === 'string') stdin = deleteOptions;
    const { file, stdout } = {
      ...((stdin ? { file: '-' } : {}) as Partial<GetOptions>),
      ...(typeof deleteOptions === 'string'
        ? ({} as Partial<GetOptions>)
        : deleteOptions)
    };
    return this.run(
      ['delete', ...(file ? ['-f', file] : [])],
      options,
      (p: ExecaChildProcess) => {
        if (stdin) {
          const stream = Readable.from([this.getStdin(stdin)]);
          if (p.stdin) stream.pipe(p.stdin);
        }
        if (stdout) {
          p.stderr?.pipe(process.stderr);
          p.stdout?.pipe(process.stdout);
        }
      }
    );
  }

  resources2String(resources: KubernetesObject[]): string {
    return resources
      .map((resource: KubernetesObject) => YAML.stringify(resource))
      .join('---\n');
  }

  string2Resources(resourcesStr: string): KubernetesObject[] {
    return `\n${resourcesStr}\n`
      .split(/\n---+\n/)
      .map(
        (resourceStr: string) =>
          YAML.parse(resourceStr.trim()) as KubernetesObject
      );
  }
}

export interface GetOptions {
  file?: string;
  ignoreNotFound?: boolean;
  output?: Output;
  stdin?: string | KubernetesObject | KubernetesObject[];
  stdout?: boolean;
}

export interface ApplyOptions {
  file?: string;
  stdin?: string | KubernetesObject | KubernetesObject[];
  stdout?: boolean;
}

export interface DeleteOptions {
  file?: string;
  stdin?: string | KubernetesObject | KubernetesObject[];
  stdout?: boolean;
}

export enum Output {
  Yaml = 'yaml',
  Json = 'json'
}

File: util/kubectl.go
/**
 * File: /util/kubectl.go
 * Project: integration-operator
 * File Created: 17-10-2023 13:49:54
 * Author: Clay Risser
 * -----
 * BitSpur (c) Copyright 2021 - 2023
 *
 * Licensed under the GNU Affero General Public License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.gnu.org/licenses/agpl-3.0.en.html
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * You can be released from the requirements of the license by purchasing
 * a commercial license. Buying such a license is mandatory as soon as you
 * develop commercial activities involving this software without disclosing
 * the source code of your own applications.
 */

package util

import (
	"context"
	"encoding/json"
	"fmt"

	"k8s.io/apimachinery/pkg/api/meta"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/client-go/discovery"
	memory "k8s.io/client-go/discovery/cached"
	"k8s.io/client-go/dynamic"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/restmapper"
	ctrl "sigs.k8s.io/controller-runtime"
)

type KubectlUtil struct {
	ctx context.Context
	cfg *rest.Config
}

func NewKubectlUtil(ctx context.Context, namespace string, serviceAccountName string) *KubectlUtil {
	cfg := ctrl.GetConfigOrDie()
	cfg.Impersonate = rest.ImpersonationConfig{
		UserName: fmt.Sprintf("system:serviceaccount:%s:%s", namespace, Default(serviceAccountName, "default")),
	}
	return &KubectlUtil{
		cfg: cfg,
		ctx: ctx,
	}
}

func (u *KubectlUtil) Create(body []byte) error {
	dr, obj, err := u.prepareDynamic(body)
	if err != nil {
		return err
	}
	if _, err := dr.Create(u.ctx, obj, metav1.CreateOptions{
		FieldManager: "integration-operator",
	}); err != nil {
		return err
	}
	return nil
}

func (u *KubectlUtil) Update(body []byte) error {
	dr, obj, err := u.prepareDynamic(body)
	if err != nil {
		return err
	}
	if _, err := dr.Update(u.ctx, obj, metav1.UpdateOptions{
		FieldManager: "integration-operator",
	}); err != nil {
		return err
	}
	return nil
}

func (u *KubectlUtil) Apply(body []byte) error {
	dr, obj, err := u.prepareDynamic(body)
	if err != nil {
		return err
	}
	data, err := json.Marshal(obj)
	if err != nil {
		return err
	}
	if _, err = dr.Patch(u.ctx, obj.GetName(), types.ApplyPatchType, data, metav1.PatchOptions{
		FieldManager: "integration-operator",
	}); err != nil {
		return err
	}
	return nil
}

func (u *KubectlUtil) Delete(body []byte) error {
	dr, obj, err := u.prepareDynamic(body)
	if err != nil {
		return err
	}
	if err = dr.Delete(u.ctx, obj.GetName(), metav1.DeleteOptions{}); err != nil {
		return err
	}
	return nil
}

func (u *KubectlUtil) Get(body []byte) (*unstructured.Unstructured, error) {
	dr, obj, err := u.prepareDynamic(body)
	if err != nil {
		return nil, err
	}
	return dr.Get(u.ctx, obj.GetName(), metav1.GetOptions{})
}

// https://ymmt2005.hatenablog.com/entry/2020/04/14/An_example_of_using_dynamic_client_of_k8s.io/client-go
func (u *KubectlUtil) prepareDynamic(resource []byte) (dynamic.ResourceInterface, *unstructured.Unstructured, error) {
	// 1. Prepare a RESTMapper to find GVR
	dc, err := discovery.NewDiscoveryClientForConfig(u.cfg)
	if err != nil {
		return nil, nil, err
	}
	mapper := restmapper.NewDeferredDiscoveryRESTMapper(memory.NewMemCacheClient(dc))

	// 2. Prepare the dynamic client
	dyn, err := dynamic.NewForConfig(u.cfg)
	if err != nil {
		return nil, nil, err
	}

	// 3. Decode YAML manifest into unstructured.Unstructured
	obj := &unstructured.Unstructured{}
	_, gvk, err := decUnstructured.Decode(resource, nil, obj)
	if err != nil {
		return nil, nil, err
	}

	// 4. Find GVR
	mapping, err := mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
	if err != nil {
		return nil, nil, err
	}

	// 5. Obtain REST interface for the GVR
	var dr dynamic.ResourceInterface
	if mapping.Scope.Name() == meta.RESTScopeNameNamespace {
		// namespaced resources should specify the namespace
		dr = dyn.Resource(mapping.Resource).Namespace(obj.GetNamespace())
	} else {
		// for cluster-wide resources
		dr = dyn.Resource(mapping.Resource)
	}

	return dr, obj, nil
}
